Source/WebCore/platform/graphics/BitmapImage.cpp

```diff
@@ -27,618 +27,143 @@
 #include "config.h"
 #include "BitmapImage.h"
 
-#include "FloatRect.h"
+#include "BitmapImageSource.h"
 #include "GeometryUtilities.h"
 #include "GraphicsContext.h"
-#include "ImageBuffer.h"
 #include "ImageObserver.h"
-#include "IntRect.h"
-#include "Logging.h"
-#include "Timer.h"
-#include <wtf/Vector.h>
-#include <wtf/text/TextStream.h>
-#include <wtf/text/WTFString.h>
-
-#if USE(CG)
-#include <pal/spi/cg/CoreGraphicsSPI.h>
-#endif
+#include "NativeImageSource.h"
 
 namespace WebCore {
 
-BitmapImage::BitmapImage(ImageObserver* observer)
-    : Image(observer)
-    , m_source(ImageSource::create(this))
-{
-}
-
-BitmapImage::BitmapImage(Ref<NativeImage>&& image)
-    : m_source(ImageSource::create(WTFMove(image)))
-{
-}
-
-BitmapImage::~BitmapImage()
+Ref<BitmapImage> BitmapImage::create(ImageObserver* observer, AlphaOption alphaOption, GammaAndColorProfileOption gammaAndColorProfileOption)
 {
-    clearTimer();
-    m_source->clearImage();
-    m_source->stopAsyncDecodingQueue();
+    return adoptRef(*new BitmapImage(observer, alphaOption, gammaAndColorProfileOption));
 }
 
-void BitmapImage::destroyDecodedData(bool destroyAll)
+Ref<BitmapImage> BitmapImage::create(Ref<NativeImage>&& nativeImage)
 {
-    LOG(Images, "BitmapImage::%s - %p - url: %s", __FUNCTION__, this, sourceURL().string().utf8().data());
-
-    if (!destroyAll) {
-        // Destroy all the frames between frame0 and m_currentFrame.
-        m_source->destroyDecodedData(1, m_currentFrame);
-    } else if (!canDestroyDecodedData()) {
-        // Destroy all the frames except frame0 and m_currentFrame.
-        m_source->destroyDecodedData(1, m_currentFrame);
-        m_source->destroyDecodedData(m_currentFrame + 1, frameCount());
-    } else {
-        m_source->destroyDecodedData(0, frameCount());
-        m_currentFrameDecodingStatus = DecodingStatus::Invalid;
-    }
-
-    // There's no need to throw away the decoder unless we're explicitly asked
-    // to destroy all of the frames.
-    if (!destroyAll || m_source->hasAsyncDecodingQueue())
-        m_source->clearFrameBufferCache(m_currentFrame);
-    else
-        m_source->resetData(data());
```

Comment:
With the code moved under BitmapImageSource::destroyDecodedData, it looks like we no longer do this. Is it unnecessary?

Reply:
No this was a mistake. I put this code back in `BitmapImageSource::destroyDecodedData()`. But I made `resetData()` just `setData()` the decoder without deleting it.